Netwayste Integration 2019

Manghi opened the pull request:
**Still a WIP**

1) Specified current versions in `Cargo.toml`. Removed binary declarations.
2) `net.rs`, `server.rs` to be used as a library. Added `lib.rs` as a result.
2) `client.rs` moved into `examples/cli-client.rs` as a command-line example
    a) Integration step will need to pull the non-user-input specific code back into a `client.rs` lib
3) Moved all tests from `net.rs`, `client.rs`, and `server.rs` into `tests.rs`. The modules specify the parent name (example `mod netwayste_client_tests { ... }`) for ease of association.

`conwayste` PR to update sub-module reference to come in a separate PR after this is completed.

Manghi committed:
Converting netwayste to library

Manghi committed:
Moved net lib tests into tests.rs

Manghi committed:
Moved server tests into tests; Moved server main() into examples/cli-server.rs

Manghi committed:
Moved client tests into tests; Moved client main() into examples/cli-client.rs

Manghi committed:
Merge branch 'aaron/dns_and_default_port' into netwaysteintegration2019

Manghi committed:
Removed cli-server and moved all code back into src/server.rs

AaronM04 commented:
Currently the `cli-client` example fails to compile:

```
$ git log -1
commit f36b0051abbd8184f249dd48afa2d6152d10ac03 (HEAD -> master, origin/master)
Author: Aaron Miller <REDACTED>
Date:   Thu Mar 28 07:29:09 2019 -0700

    submodule bump
$ ls
Cargo.lock  Cargo.toml  conwayste/  conwayste.toml  libconway/  LICENSE.txt  netwayste/  README.md  target/
$ cargo run --example cli-client chococat.conwayste.rs
   Compiling netwayste v0.1.1 (/REDACTED/conwayste/netwayste)
error[E0583]: file not found for module `net`
  --> netwayste/examples/cli-client.rs:31:5
   |
31 | mod net;
   |     ^^^
   |
   = help: name the file either net.rs or net/mod.rs inside the directory "netwayste/examples"

error: aborting due to previous error

For more information about this error, try `rustc --explain E0583`.
error: Could not compile `netwayste`.

To learn more, run the command again with --verbose.
$ cd netwayste
$ git log -1
commit 56489739483050b38bd8ceea7969bf9098a9f02f (HEAD -> netwaysteintegration2019, origin/netwaysteintegration2019)
Author: manghi <REDACTED>
Date:   Mon Mar 25 22:52:29 2019 -0700

    Removed cli-server and moved all code back into src/server.rs
```

AaronM04 commented:
Deleted my previous comment because I had copied another file over `netwayste/examples/cli-client.rs` :man_facepalming:

AaronM04 commented on examples/cli-client.rs:
```diff
@@ -0,0 +1,245 @@
+#[macro_use]
```
Please add copyright block above this line :)

AaronM04 commented on src/server.rs:
```diff
@@ -1191,15 +1176,15 @@ fn main() {
         });
 
-    trace!("Listening for connections on {:?}...", udp.local_addr().unwrap());
+     trace!("Listening for connections on {:?}...", udp.local_addr().unwrap());
```
extra space at start of line

AaronM04 commented on src/server.rs:
```diff
@@ -52,27 +34,29 @@ use std::collections::HashMap;
+use crate::semver::Version;
+use crate::clap::{App, Arg};
+
+pub const TICK_INTERVAL_IN_MS:    u64      = 10;
```
Any particular reason to make the stuff in `server.rs` public? We will only have one server, right? I was thinking this `server.rs` could be our permanent server program, rather than being imported elsewhere.

Manghi replied:
Was thinking about if we ever wanted a gui-based server controller.
[See an example here](https://imgur.com/qDJplm8)

I'll undo it for now since this wasn't a complicated refactor.

AaronM04 reviewed (commented):
Good work on consolidating the tests! :grinning:

AaronM04 commented on Cargo.toml:
```diff
@@ -4,28 +4,20 @@ version = "0.1.1"
 
-[[bin]]
-name = "server"
-path = "src/server.rs"
```
How do I start the server?

AaronM04 commented on examples/cli-server.rs:
```diff
@@ -0,0 +1,193 @@
+#[macro_use]
```
Just noticed this... see my other comment about `server.rs` being the permanent thing.

Manghi committed:
Addressed initial pass of code review comments

Manghi committed:
Added start_network to ClientState

Manghi committed:
Moved address into client state

Manghi committed:
[WIP] Working on a channel linkage between netwayste and conwayste

Commit:
HIGHLY SUSPECT COMMIT -- HELLA WIP

Commit:
highly suspect commit's little brother WIP

Manghi committed:
tiny bit of cleaning

Manghi committed:
Version bump to 0.2.0

Manghi committed:
Moving logging initialization to conwayste

Manghi committed:
Added NetwaysteEvent and some minor clean-up

Manghi commented:
@AaronM04 Happy with the basic integration that's been reached. I would like to discuss more on the next few steps before ripping apart the conwayste code as its in dire need of reconstruction.

Biggest change since the last review? I replaced `ResponseCode` and `RequestAction` on the channel ends with a new enumerator, `NetwaysteEvent` so that a single type is sent and received on the channels. It's members are similar to a request/response enums, but that is because I did not want to send xfer packets through the channel. (There's info in a packet that the Conwayste layer should not ever care about.)

Ready for review!

Manghi committed:
Updated example to use NetwaysteEvent

Manghi committed:
Fixed ClientNetState tests

Manghi committed:
[DNC] Fixed netwayste client tests; server.rs relocation

- Updated serde to use "edition 2018" updates
- Netwayste client tests updated to compile, needs fixing
- Server.rs moved out of bin into main project root
    - tests cannot find the module...working on it

AaronM04 commented on examples/cli-client.rs:
```diff
@@ -0,0 +1,61 @@
+        loop {
+            match ggez_server_response.try_recv() {
+                Ok(response_code) => {
+                    println!("sweet! we got a netwayste ResponseCode! {:?}", response_code);
```
Can I get the original CLI client back? It was very useful :)

Manghi replied:
Done! I think a little bit of touch up on the output end is needed when #25 is implemented

Manghi committed:
Moved server tests into server.rs; tests pass

Manghi committed:
Remove unnecessary 'as ...'

Manghi commented:
@AaronM04  Fixed a couple things, of which I moved `src\bin\server.rs` (I didn't like how I put it there in the first place) to `src\server.rs`. I had to move the server tests back into `src\server.rs` because I couldn't find the server module in the tests server-specific mod and I spent far too long trying various things out. You're free to give it a shot but you've been warned. At least the tests are now passing.

No updates related to your comments here or in conwayste/conwayste/pull/30; will get to those another night.

Manghi committed:
Ported historic netwayste cli to cli-client.rs

Manghi committed:
Specified receive error disconnect case

AaronM04 commented on src/net.rs:
```diff
@@ -337,6 +337,7 @@ impl Ord for Packet {
 }
 
 //////////////// Packet (de)serialization ////////////////
+#[allow(dead_code)]
```
I'm not a huge fan of this. If we don't use it, we should take it out.

AaronM04 commented on src/net.rs:
```diff
@@ -942,14 +944,14 @@ impl NetworkManager {
 }
 
 #[derive(PartialEq, Debug, Clone)]
+#[allow(dead_code)]
```
Why do we need this on an `enum`?

Commit:
[WIP] cli-client reworking, and change client keepalive logic a bit; need to fix cli-client tests

Commit:
nw client: set name in ClientNetState on NetwaysteEvent::Connect

Also, use 33ms sleep in cli-client example, and color-backtrace.

AaronM04 commented:
Thanks! Merging now :rocket:
